packages/core/src/behavior.ts

```diff
@@ -0,0 +1,359 @@
+  let dispose;
+
+  const behavior: Behavior<TEvent, undefined> = {
+    receive: (_, event) => {
+      const plainEvent = isSCXMLEvent(event) ? event.data : event;
+      receivers.forEach((receiver) => receiver(plainEvent));
+
+      return behavior;
+    },
+    receiveSignal: (actorContext, signal) => {
+      if (signal === startSignal) {
+        dispose = callback(
+          (e) => {
+            if (canceled) {
+              return;
+            }
+
+            parent?.send(toSCXMLEvent(e, { origin: actorContext.self }));
+          },
+          (newListener) => {
+            receivers.add(newListener);
+          }
+        );
+
+        if (isPromiseLike(dispose)) {
+          dispose.then(
+            (resolved) => {
+              parent?.send(
+                toSCXMLEvent(doneInvoke(actorContext.name, resolved) as any, {
+                  origin: actorContext.self
+                })
+              );
+              canceled = true;
+            },
+            (errorData) => {
+              const errorEvent = error(actorContext.name, errorData);
+              parent?.send(
+                toSCXMLEvent(errorEvent, { origin: actorContext.self })
+              );
+              canceled = true;
+            }
+          );
+        }
+      }
+
+      if (signal === stopSignal) {
+        canceled = true;
+
+        if (isFunction(dispose)) {
+          dispose();
+        }
+      }
+
+      return behavior;
+    },
+    current: undefined
+  };
+
+  return behavior;
+}
+
+export function createPromiseBehavior<T, TEvent extends EventObject>(
+  promise: PromiseLike<T>,
+  parent?: ActorRef<any>
+): Behavior<TEvent, T | undefined> {
+  let canceled = false;
+  const observers: Set<Observer<T>> = new Set();
+
+  const behavior: Behavior<TEvent, T | undefined> = {
+    receive: () => {
+      return behavior;
+    },
+    receiveSignal: (actorContext: ActorContext, signal: LifecycleSignal) => {
+      switch (signal) {
+        case startSignal:
+          const resolvedPromise = Promise.resolve(promise);
+
+          resolvedPromise.then(
+            (response) => {
+              if (!canceled) {
+                parent?.send(
+                  toSCXMLEvent(doneInvoke(actorContext.name, response) as any, {
+                    origin: actorContext.self
+                  })
+                );
+
+                observers.forEach((observer) => {
+                  observer.next?.(response);
+                  observer.complete?.();
+                });
+              }
+            },
+            (errorData) => {
+              if (!canceled) {
+                const errorEvent = error(actorContext.name, errorData);
+
+                parent?.send(
+                  toSCXMLEvent(errorEvent, { origin: actorContext.self })
+                );
+
+                observers.forEach((observer) => {
+                  observer.error?.(errorData);
+                  observer.complete?.();
```

Comment:
this doesnt comply to common observable implementations, error and complete callbacks are mutually exclusive (only one of them can be called), see an example here: https://codesandbox.io/s/nice-water-7dfhu?file=/src/index.js